src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp

```diff
@@ -3177,4 +3290,974 @@ void StubGenerator::aesgcm_encrypt(Register in, Register len, Register ct, Regis
+  __ vpxor(t7, t7, t4, Assembler::AVX_128bit);
+
+  __ vpclmulqdq(t2, t2, t3, 0x00);
+
+  __ vpxor(xmm1, xmm1, t2, Assembler::AVX_128bit);
+
+  __ movdqu(t5, Address(subkeyHtbl, 1 * 16));
+  __ vpshufd(t2, xmm8, 78, Assembler::AVX_128bit);
+  __ vpshufd(t3, t5, 78, Assembler::AVX_128bit);
+  __ vpxor(t2, t2, xmm8, Assembler::AVX_128bit);
+  __ vpxor(t3, t3, t5, Assembler::AVX_128bit);
+
+  __ vpclmulqdq(t4, xmm8, t5, 0x11);
+  __ vpxor(t6, t6, t4, Assembler::AVX_128bit);
+
+  __ vpclmulqdq(t4, xmm8, t5, 0x00);
+  __ vpxor(t7, t7, t4, Assembler::AVX_128bit);
+
+  __ vpclmulqdq(t2, t2, t3, 0x00);
+
+  __ vpxor(xmm1, xmm1, t2, Assembler::AVX_128bit);
+  __ vpxor(xmm1, xmm1, t6, Assembler::AVX_128bit);
+  __ vpxor(t2, xmm1, t7, Assembler::AVX_128bit);
+
+  __ vpslldq(t4, t2, 8, Assembler::AVX_128bit);
+  __ vpsrldq(t2, t2, 8, Assembler::AVX_128bit);
+
+  __ vpxor(t7, t7, t4, Assembler::AVX_128bit);
+  __ vpxor(t6, t6, t2, Assembler::AVX_128bit); //<t6:t7> holds the result of the accumulated carry-less multiplications
+
+  //first phase of the reduction
+  __ movdqu(t3, ExternalAddress(ghash_polynomial_reduction_addr()), rbx /*rscratch*/);
+
+  __ vpclmulqdq(t2, t3, t7, 0x01);
+  __ vpslldq(t2, t2, 8, Assembler::AVX_128bit); // shift-L t2 2 DWs
+
+  __ vpxor(t7, t7, t2, Assembler::AVX_128bit);//first phase of the reduction complete
+
+  //second phase of the reduction
+  __ vpclmulqdq(t2, t3, t7, 0x00);
+  __ vpsrldq(t2, t2, 4, Assembler::AVX_128bit); //shift-R t2 1 DW (Shift-R only 1-DW to obtain 2-DWs shift-R)
+
+  __ vpclmulqdq(t4, t3, t7, 0x10);
+  __ vpslldq(t4, t4, 4, Assembler::AVX_128bit); //shift-L t4 1 DW (Shift-L 1-DW to obtain result with no shifts)
+  __ vpxor(t4, t4, t2, Assembler::AVX_128bit); //second phase of the reduction complete
+  __ vpxor(t6, t6, t4, Assembler::AVX_128bit); //the result is in t6
+}
+
+void StubGenerator::initial_blocks(XMMRegister ctr, Register rounds, Register key,
+  Register len, Register in, Register out, Register ct, Register subkeyHtbl, Register pos) {
+  const XMMRegister t1 = xmm12;
+  const XMMRegister t2 = xmm13;
+  const XMMRegister t3 = xmm14;
+  const XMMRegister t4 = xmm15;
+  const XMMRegister t5 = xmm11;
+  const XMMRegister t6 = xmm10;
+  const XMMRegister t_key = xmm0;
+
+  Label skip_reload, last_aes_rnd, aes_192, aes_256;
+  //Move AAD_HASH to temp reg
+  __ movdqu(t2, xmm8);
+  //Prepare 8 counter blocks and perform rounds of AES cipher on
+  //them, load plain/cipher text and store cipher/plain text.
+  __ movdqu(xmm1, xmm9);
+  __ movdqu(t5, ExternalAddress(counter_mask_linc1_addr()), rbx /*rscratch*/);
+  __ vpaddd(xmm2, xmm1, t5, Assembler::AVX_128bit);
+  __ vpaddd(xmm3, xmm2, t5, Assembler::AVX_128bit);
+  __ vpaddd(xmm4, xmm3, t5, Assembler::AVX_128bit);
+  __ vpaddd(xmm5, xmm4, t5, Assembler::AVX_128bit);
+  __ vpaddd(xmm6, xmm5, t5, Assembler::AVX_128bit);
+  __ vpaddd(xmm7, xmm6, t5, Assembler::AVX_128bit);
+  __ vpaddd(xmm8, xmm7, t5, Assembler::AVX_128bit);
```

Comment:
This could be done more efficiently as follows:
  __ movdqu(t5, ExternalAddress(counter_mask_linc1_addr()), rbx /*rscratch*/);
  __ movdqu(t6, ExternalAddress(counter_mask_linc2_addr()), rbx /*rscratch*/);
  __ vpaddd(xmm2, xmm1, t5, Assembler::AVX_128bit);
  __ vpaddd(xmm3, xmm1, t6, Assembler::AVX_128bit);
  __ vpaddd(xmm4, xmm2, t6, Assembler::AVX_128bit);
  __ vpaddd(xmm5, xmm3, t6, Assembler::AVX_128bit);
  __ vpaddd(xmm6, xmm4, t6, Assembler::AVX_128bit);
  __ vpaddd(xmm7, xmm5, t6, Assembler::AVX_128bit);
  __ vpaddd(xmm8, xmm6, t6, Assembler::AVX_128bit);